File: plugins/dev/skills/publish-artifact/scripts/parse_target.py
#!/usr/bin/env python3
"""Parse a PR/issue target (bare number or GitHub URL) into structured fields.

Pure: no network. A bare number yields kind="unknown" — the caller probes
whether it is a PR or an issue. A URL yields kind from its /pull/ or /issues/ path.
"""
import argparse
import json
import re
import sys

_URL_RE = re.compile(
    r"^https?://github\.com/(?P<owner>[^/]+)/(?P<repo>[^/]+)/(?P<kind>pull|issues)/(?P<number>\d+)"
)


def parse(target, current_repo):
    """Return {owner, repo, number:int, kind:'pr'|'issue'|'unknown'}.

    current_repo is "owner/repo" for bare-number targets. Raises ValueError on
    unrecognizable input or a bare number with no current_repo.
    """
    m = _URL_RE.match(target.strip())
    if m:
        return {
            "owner": m.group("owner"),
            "repo": m.group("repo"),
            "number": int(m.group("number")),
            "kind": "pr" if m.group("kind") == "pull" else "issue",
        }
    if re.fullmatch(r"\d+", target.strip()):
        if "/" not in current_repo:
            raise ValueError("bare number requires a current repo (owner/repo)")
        owner, repo = current_repo.split("/", 1)
        return {"owner": owner, "repo": repo, "number": int(target), "kind": "unknown"}
    raise ValueError(f"Target must be a PR/issue number or a GitHub URL: {target!r}")


def main(argv=None):
    ap = argparse.ArgumentParser()
    ap.add_argument("target")
    ap.add_argument("--current-repo", default="", help='"owner/repo" for bare-number targets')
    args = ap.parse_args(argv)
    try:
        print(json.dumps(parse(args.target, args.current_repo)))
        return 0
    except ValueError as e:
        print(str(e), file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: cli/src/agent-tools/github-pr/parse-target.ts
/**
 * Pure parsing of a PR/issue target into structured fields.
 *
 * A bare number yields `kind: "unknown"` — the caller probes whether it is a PR
 * or an issue. A GitHub URL yields `kind` from its `/pull/` or `/issues/` path.
 */

/** Whether a parsed target is known to be a PR, an issue, or undetermined. */
export type TargetKind = "pr" | "issue" | "unknown";

/** A target resolved into the repo coordinates needed to address it. */
export interface ParsedTarget {
	readonly owner: string;
	readonly repo: string;
	readonly number: number;
	readonly kind: TargetKind;
}

/**
 * A GitHub PR/issue URL: owner, repo, the `pull`/`issues` segment, and number.
 * Trailing path segments (e.g. `/files`) are ignored.
 */
const URL_RE =
	/^https?:\/\/github\.com\/(?<owner>[^/]+)\/(?<repo>[^/]+)\/(?<kind>pull|issues)\/(?<number>\d+)(?=$|[/?#])/;

/** A target consisting solely of digits. */
const BARE_NUMBER_RE = /^\d+$/;

/**
 * Parse a target into `{owner, repo, number, kind}`.
 *
 * `currentRepo` is `"owner/repo"`, used only for bare-number targets. Throws
 * when the input is neither a recognizable GitHub URL nor a bare number, or
 * when a bare number is given without a valid `currentRepo`.
 *
 * @param target - A bare PR/issue number or a GitHub PR/issue URL.
 * @param currentRepo - `"owner/repo"` context for bare-number targets.
 */
export const parse = (target: string, currentRepo: string): ParsedTarget => {
	const trimmed = target.trim();
	const m = URL_RE.exec(trimmed);
	if (m?.groups) {
		return {
			owner: m.groups.owner,
			repo: m.groups.repo,
			number: Number.parseInt(m.groups.number, 10),
			kind: m.groups.kind === "pull" ? "pr" : "issue",
		};
	}
	if (BARE_NUMBER_RE.test(trimmed)) {
		if (!currentRepo.includes("/")) {
			throw new Error("bare number requires a current repo (owner/repo)");
		}
		const slash = currentRepo.indexOf("/");
		return {
			owner: currentRepo.slice(0, slash),
			repo: currentRepo.slice(slash + 1),
			number: Number.parseInt(trimmed, 10),
			kind: "unknown",
		};
	}
	throw new Error(
		`Target must be a PR/issue number or a GitHub URL: ${JSON.stringify(target)}`,
	);
};
